src/main/kotlin/com/jetbrains/micropython/repl/ToolWindowReplTab.kt

```diff
@@ -0,0 +1,120 @@
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.DefaultActionGroup
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.ModalityState
+import com.intellij.openapi.project.DumbAware
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Disposer
+import com.jediterm.terminal.TtyConnector
+import com.jetbrains.micropython.settings.MicroPythonDevicesConfiguration
+import org.jetbrains.plugins.terminal.JBTerminalSystemSettingsProvider
+import org.jetbrains.plugins.terminal.ShellTerminalWidget
+import java.awt.BorderLayout
+import javax.swing.JPanel
+
+class ToolWindowReplTab(val project: Project, parent: Disposable) : CommsEventObserver {
+    val deviceCommsManager = DeviceCommsManager.getInstance(project)
+    private val deviceConfiguration = MicroPythonDevicesConfiguration.getInstance(project)
+    val terminalWidget: ShellTerminalWidget
+
+    init {
+        val mySettingsProvider = JBTerminalSystemSettingsProvider()
+        terminalWidget = ShellTerminalWidget(project, mySettingsProvider, parent)
+
+        Disposer.register(terminalWidget, mySettingsProvider)
+
+        deviceCommsManager.registerObserver(this)
+        if (!deviceCommsManager.isRunning()) {
+            deviceCommsManager.startREPL()
+        }
+    }
+
+    private fun connectWidgetTty(
+        terminalWidget: ShellTerminalWidget,
+        connector: TtyConnector
+    ) {
+        if (terminalWidget.isSessionRunning) {
+            terminalWidget.stop()
+        }
+        terminalWidget.start(connector)
+        val modalityState = ModalityState.stateForComponent(terminalWidget.component)
+
+        ApplicationManager.getApplication().invokeLater(
+            {
+                try {
+                    terminalWidget.component.revalidate()
+                    terminalWidget.notifyStarted()
+                } catch (e: RuntimeException) {
+                    TODO("You can't cut back on error reporting! You will regret this!")
+                }
+            },
+            modalityState
+        )
+    }
+
+    override fun onCommsEvent(event: CommsEvent) {
+        when (event) {
+            is CommsEvent.ProcessStarted -> {
+                if (deviceConfiguration.clearReplOnLaunch) {
+                    terminalWidget.terminal.clearScreen()
+                } else {
+                    terminalWidget.terminal.nextLine()
+                }
+                connectWidgetTty(terminalWidget, event.ttyConnector)
+            }
+            is CommsEvent.ProcessCreationFailed -> {
+                terminalWidget.terminal.nextLine()
+                terminalWidget.currentSession.terminal.writeCharacters(event.reason)
+            }
+            CommsEvent.ProcessDestroyed -> terminalWidget.stop()
+        }
+    }
+
+    fun createUI(): JPanel {
+        val toolbarActions = DefaultActionGroup()
```

Comment:
You can use `toolbarActions.apply { add(...) }` here as you did below with the `JPanel`.